File: hs/src/Day21.hs
{-|
Module:         Day21
Description:    <https://adventofcode.com/2023/day/21 Day 21: Step Counter>
-}
{-# LANGUAGE ViewPatterns #-}
module Day21 (part1, part2) where

import Control.Monad (join)
import Data.Bits ((.&.), (.^.))
import Data.Bool (bool)
import Data.Heap (FstMinPolicy)
import qualified Data.Heap as Heap (empty, insert, view)
import Data.List (foldl')
import Data.Map (Map)
import qualified Data.Map as Map ((!), (!?), empty, insert, intersectionWith, notMember)
import Data.Monoid (Sum(Sum), getSum)
import qualified Data.Set as Set (fromDistinctAscList, member, toList)
import Data.Text (Text)
import qualified Data.Text as T (all, head, index, last, length, lines)
import Data.Tuple (swap)
import qualified Data.Vector as V ((!), all, fromList, head, last, length)

bfs :: (Ord a, Enum b, Ord b) => (a -> [a]) -> [(a, b)] -> Map a b
bfs neighbors = bfs' Map.empty . foldl' (flip $ Heap.insert @FstMinPolicy . swap) Heap.empty where
    bfs' visited (Heap.view -> Just ((b, a), q))
      | Just c <- visited Map.!? a = if c > b then error "invalid state" else bfs' visited q
      | otherwise = bfs' (Map.insert a b visited) . foldl' (flip $ Heap.insert . (succ b,)) q .
            filter (`Map.notMember` visited) $ neighbors a
    bfs' visited _ = visited

part1 :: Int -> Text -> Int
part1 n input = getSum $ foldMap (Sum . bool 0 1 . ok) distances where
    grid = V.fromList $ T.lines input
    keys = Set.fromDistinctAscList
      [ (y, x)
      | y <- [0..V.length grid - 1]
      , let line = grid V.! y
      , x <- [0..T.length line - 1]
      , line `T.index` x /= '#'
      ]
    neighbors (y, x) = filter (`Set.member` keys) [(y - 1, x), (y, x - 1), (y, x + 1), (y + 1, x)]
    distances = bfs neighbors
        [(p, 0 :: Int) | p@(y, x) <- Set.toList keys, grid V.! y `T.index` x == 'S']
    ok x = x <= n && even (x .^. n)

part2 :: Int -> Text -> Int
part2 n input
  | odd size, V.all ((== size) . T.length) grid
  , T.all (/= '#') $ V.head grid, T.all (/= '#') $ V.last grid
  , V.all (\line -> T.head line /= '#' && T.last line /= '#') grid
  = getSum $ foldMap (Sum . countOne) origin <>
        foldMap (foldMap $ Sum . countQuadrant) quadrants <>
        mconcat (axes >>= countAxis)
  where
    grid = V.fromList $ T.lines input
    size = V.length grid
    keys = Set.fromDistinctAscList
      [ (y, x)
      | y <- [0..V.length grid - 1]
      , let line = grid V.! y
      , x <- [0..T.length line - 1]
      , line `T.index` x /= '#'
      ]
    neighbors (y, x) = filter (`Set.member` keys) [(y - 1, x), (y, x - 1), (y, x + 1), (y + 1, x)]

    origin = bfs neighbors
        [(p, 0 :: Int) | p@(y, x) <- Set.toList keys, grid V.! y `T.index` x == 'S']
    countOne x = if x <= n && even (x .^. n) then 1 else 0

    quadrants =
      [ bfs neighbors [((y, x), origin Map.! (size - 1 - y, size - 1 - x) + 2)]
      | y <- [0, size - 1], x <- [0, size - 1]
      ]
    countQuadrant x
      | x > n = 0
      | even r = join (*) $ r `div` size `div` 2 + 1
      | otherwise = (succ >>= (*)) $ (r `div` size + 1) `div` 2
      where r = n - x

    axes =
      [ iterate (bfs neighbors . copy) origin
      | f <- [(,), flip (,)], s <- [0, size - 1]
      , let copy distances = [(f s t, distances Map.! f (size - 1 - s) t + 1) | t <- [0..size - 1]]
      ]
    countAxis axis = foldMap (Sum . countRest) rest : (foldMap (Sum . countOne) <$> blocks) where
        done (prev, cur) = all (> n) cur || all (== size) (Map.intersectionWith (-) cur prev)
        (map snd -> blocks, (_, rest):_) = break done . zip axis $ tail axis
        countRest x
          | x > n = 0
          | otherwise = max 0 $ (r `div` size - (r .&. 1)) `div` 2 + 1
          where r = n - x

File: py/aoc2023/day21.py
"""
Day 21: Step Counter
"""

SAMPLE_INPUT = """
...........
.....###.#.
.###.##..#.
..#.#...#..
....#.#....
.##..S####.
.##..#...#.
.......##..
.##.#.####.
.##..##.##.
...........
"""


def _count(grid, start, n):
    frontier, visited, acc = {start}, set(), 0
    for d in range(n):
        if not (d ^ n) & 1:
            acc += len(frontier)
        visited |= frontier
        frontier = {
            (y1, x1)
            for y0, x0 in frontier
            for y1, x1 in [(y0 - 1, x0), (y0, x0 - 1), (y0, x0 + 1), (y0 + 1, x0)]
            if 0 <= y1 < len(grid)
            and 0 <= x1 < len(grid[y1])
            and grid[y1][x1] != "#"
            and (y1, x1) not in visited
        }
    return acc + len(frontier)


def part1(data, n=64):
    """
    >>> part1(SAMPLE_INPUT, n=1)
    2
    >>> part1(SAMPLE_INPUT, n=2)
    4
    >>> part1(SAMPLE_INPUT, n=3)
    6
    >>> part1(SAMPLE_INPUT, n=6)
    16
    """
    grid = [line for line in data.splitlines() if line]
    (start,) = (
        (y, x) for y, line in enumerate(grid) for x, c in enumerate(line) if c == "S"
    )
    return _count(grid, start, n)


def part2(data, n=26501365):
    """
    >>> part2(SAMPLE_INPUT, n=6) # doctest: +SKIP
    16
    >>> part2(SAMPLE_INPUT, n=10) # doctest: +SKIP
    50
    >>> part2(SAMPLE_INPUT, n=50) # doctest: +SKIP
    1594
    >>> part2(SAMPLE_INPUT, n=100) # doctest: +SKIP
    6536
    >>> part2(SAMPLE_INPUT, n=500) # doctest: +SKIP
    167004
    >>> part2(SAMPLE_INPUT, n=1000) # doctest: +SKIP
    668697
    >>> part2(SAMPLE_INPUT, n=5000) # doctest: +SKIP
    16733044
    """
    grid = [line for line in data.splitlines() if line]
    m = len(grid)
    q, r = n // m, n % m
    ((y0, x0),) = (
        (y, x) for y, line in enumerate(grid) for x, c in enumerate(line) if c == "S"
    )
    a, b, c, d = (
        _count(
            [line * (2 * i + 1) for line in grid] * (2 * i + 1),
            (y0 + i * m, x0 + i * m),
            r + i * m,
        )
        for i in range(4)
    )
    assert d == a - 3 * b + 3 * c
    return a + (b - a) * q + (c - 2 * b + a) * (q * (q - 1) // 2)


parts = (part1, part2)
